### PR TITLE
Problem: sender address verification is not decoupled from execution

### PR DESCRIPTION
**Pull request by yihuang**
## Description

WIP: #355

Solution:
- to decouple sender address verification from execution, we must first enforce user setting the From field in the msg
- change `From` field from string to bytes, avoid heavy checksum hex encoding of go-ethereum
- also minimize checksum hex encoding in other places in state machine.

## Benchmark Results

Run the `BenchmarkERC20Transfer` in cronos repo.

```
goos: darwin
goarch: amd64
pkg: github.com/crypto-org-chain/cronos/v2/app
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
                         │ /tmp/after.txt │           /tmp/after2.txt           │
                         │     sec/op     │    sec/op     vs base               │
ERC20Transfer/memdb-12       612.9m ± 19%   602.1m ± 12%        ~ (p=0.818 n=6)
ERC20Transfer/leveldb-12     700.5m ± 18%   593.4m ±  1%  -15.28% (p=0.002 n=6)
ERC20Transfer/memiavl-12     604.1m ± 28%   582.6m ± 23%        ~ (p=0.240 n=6)
geomean                      637.7m         592.6m         -7.07%

                         │ /tmp/after.txt │          /tmp/after2.txt           │
                         │      B/op      │     B/op      vs base              │
ERC20Transfer/memdb-12       176.4Mi ± 0%   174.1Mi ± 0%  -1.32% (p=0.002 n=6)
ERC20Transfer/leveldb-12     178.2Mi ± 0%   175.9Mi ± 0%  -1.30% (p=0.002 n=6)
ERC20Transfer/memiavl-12     171.0Mi ± 0%   168.7Mi ± 0%  -1.36% (p=0.002 n=6)
geomean                      175.2Mi        172.9Mi       -1.33%

                         │ /tmp/after.txt │          /tmp/after2.txt          │
                         │   allocs/op    │  allocs/op   vs base              │
ERC20Transfer/memdb-12        2.802M ± 0%   2.776M ± 0%  -0.93% (p=0.002 n=6)
ERC20Transfer/leveldb-12      2.801M ± 0%   2.775M ± 0%  -0.93% (p=0.002 n=6)
ERC20Transfer/memiavl-12      2.747M ± 0%   2.721M ± 0%  -0.95% (p=0.002 n=6)
geomean                       2.783M        2.757M       -0.93%
```

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: app/ante/setup.go [hunk 1]
```diff
@@ -163,11 +163,6 @@ func (vbd EthValidateBasicDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simu
 			return ctx, errorsmod.Wrapf(errortypes.ErrUnknownRequest, "invalid message type %T, expected %T", msg, (*evmtypes.MsgEthereumTx)(nil))
 		}
 
-		// Validate `From` field
-		if msgEthTx.From != "" {
```
Comment (yihuang) [inline app/ante/setup.go:167]: moved to `ValidateBasic`.

File: x/evm/types/utils.go [hunk 2]
```diff
@@ -181,3 +182,11 @@ func BinSearch(lo, hi uint64, executable func(uint64) (bool, *MsgEthereumTxRespo
 func EffectiveGasPrice(baseFee *big.Int, feeCap *big.Int, tipCap *big.Int) *big.Int {
 	return math.BigMin(new(big.Int).Add(tipCap, baseFee), feeCap)
 }
+
+// HexAddress encode ethereum address without checksum, faster to run for state machine
+func HexAddress(a []byte) string {
```
Comment (mmsqe) [inline x/evm/types/utils.go:187]: do we add any check for HexAddress

Comment (yihuang) [reply x/evm/types/utils.go:187]: just hex encoding, what kind of check?

Comment (yihuang) [reply x/evm/types/utils.go:187]: it's used by event emitting, we shouldn't deal with hex stuff in state machine in the first place.
it's to replace the checksum hex method provided by go-ethereum.

**Review by mmsqe — DISMISSED** [review]

File: store/cachekv/internal/memiterator.go [hunk 3]
```diff
@@ -23,6 +23,10 @@ type memIterator struct {
 }
 
 func newMemIterator(start, end []byte, items BTree, ascending bool) *memIterator {
+	if items.tree == nil {
+		return &memIterator{start: start, end: end, ascending: ascending, valid: false}
```
Comment (mmsqe) [inline store/cachekv/internal/memiterator.go:27]: not sure why valid false here

Comment (yihuang) [reply store/cachekv/internal/memiterator.go:27]: `tree == nil` means the tree is empty, so any iterator simply returns nothing.
it's for saving some allocations when creating cache store wrappers, where most of the stores are not used at all.

**Review by mmsqe — APPROVED** [review]

**Pull request merged**